com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs

```diff
@@ -212,14 +212,23 @@ public void OnProcessShader(Shader shader, ShaderSnippetData snippetData, IList<
                 return;
 
             int prevVariantCount = compilerDataList.Count;
-
-            for (int i = 0; i < compilerDataList.Count; ++i)
+            
+            var inputShaderVariantCount = compilerDataList.Count;
+            for (int i = 0; i < inputShaderVariantCount;)
             {
-                if (StripUnused(ShaderBuildPreprocessor.supportedFeatures, shader, snippetData, compilerDataList[i]))
-                {
+                bool removeInput = StripUnused(ShaderBuildPreprocessor.supportedFeatures, shader, snippetData, compilerDataList[i]);
+                if (removeInput)
+                    compilerDataList[i] = compilerDataList[--inputShaderVariantCount];
+                else
```

Comment:
So, in this case we need to remove compilerDataList[i]. however removing here is a O(n) operation. so instead of removing we overwrite it with the last element and decrement the last element count. 

The code above is same as:
if (removeInput)
{
    compilerDataList[i] = compilerDataList[inputShaderVariantCount];
    inputShaderVariantCount--;
}

Reply:
ah ok